webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,219 @@
+    isLoading?: (b: boolean) => void,
+  ) {
+    this.#wsUrl = wsUrl;
+    this.#onUpsert = onUpsert;
+    this.#onDelete = onDelete;
+    this.#isLoading = isLoading;
+    this.#advance();
+  }
+
+  #connect() {
+    this.#ws = new WebSocket(this.#wsUrl);
+    this.#ws.onopen = () => {
+      logger.info('Streaming websocket opened!');
+      this.#retries = 0;
+      this.#advance();
+    };
+    this.#ws.onerror = (err) => {
+      // No need to do anything else becauses onerror will trigger onclose
+      logger.error('Streaming websocket errored: ', err);
+    };
+
+    this.#ws.onclose = () => {
+      this.#syncSent = undefined;
+      const backoff = backoffs[this.#retries];
+      if (backoff === undefined) {
+        throw new Error('Websocket cannot reconnect!');
+      }
+      this.#timeout = dayjs().add(backoff, 'second');
+      this.#retries += 1;
+      logger.info(`#${this.#retries} of retries: in ${backoff}s`);
+      setTimeout(() => this.#advance(), backoff * 1000);
+    };
+
+    this.#ws.onmessage = (event) => {
+      const msg = JSON.parse(event.data) as Record<string, StreamContent>;
+      this.#pendingMsg.push(msg);
+      this.#advance();
+    };
+  }
+
+  #shouldSkip(newSub: Record<Streamable, Subscription>): boolean {
+    if (!this.#curSub) return false;
+    let same = true;
+    forEach(newSub, (val, k) => {
+      if (!this.#curSub?.[k as Streamable].spec.equals(val.spec)) same = false;
+    });
+    return same;
+  }
+
+  #sendSpec(newSub?: Record<Streamable, Subscription>): void {
+    if (!newSub) return;
+    // Skip current sub and move to next only if sub has already been sent
+    if (this.#shouldSkip(newSub) && this.#syncSent) {
+      this.#advance();
+      return;
+    }
+
+    this.#numSyncs += 1;
+    const sync_id = this.#numSyncs.toString();
+    const spec = reduce(
+      newSub,
+      (spec, ent, k) => {
+        spec['known'][k] = ent.keyCache.known();
+        spec['subscribe'][k] = {
+          ...ent.spec.toWire(),
+          since: ent.keyCache.maxSeq(),
+        };
+        return spec;
+      },
+      { known: {}, subscribe: {}, sync_id: sync_id } as Record<string, StreamContent>,
+    );
+
+    this.#curSub = newSub;
+    this.#ws && this.#ws.send(JSON.stringify(spec));
+    this.#syncSent = sync_id;
+  }
+
+  #processMsg(): void {
+    while (this.#pendingMsg.length > 0) {
+      const msg = this.#pendingMsg.shift();
+      if (!msg) break;
+      if (msg['sync_id']) {
+        if (!msg['complete']) {
+          this.#syncStarted = msg['sync_id'];
+          this.#isLoading?.(false);
+        } else {
+          this.#syncComplete = msg['sync_id'];
+          this.#isLoading?.(true);
+        }
+      } else if (this.#syncSent === this.#syncStarted) {
+        // Ignore all messages between when we send a new subscription and when the
+        // sync-start message for that subscription arrives.  These are the online
+        // updates for a subscription we no longer care about.
```

Comment:
ok, I see you did take my advice about using an else-if, but the way you did it left this comment in a state where the comment only applies when it's _not_ true.  Like, the comment is talking about the case where `this.#syncSent !== this.#syncStarted` but it's in a block that runs in the opposite case.  I had this in mind:

```ts
} else if (this.#syncSent !== this.#syncStarted) {
  // Ignore all messages between when we send a new subscription and when the
  // sync-start message for that subscription arrives.  These are the online
  // updates for a subscription we no longer care about.
} else {
  forEach(
     ...
  );
}
```